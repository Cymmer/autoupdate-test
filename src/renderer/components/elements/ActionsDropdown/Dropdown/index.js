import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import styles from './styles.module.scss';
import Card from '../../Card';

import actionTypes from '../constants/actionTypes';
import ActionLink from '../Action/Link';
import ActionButton from '../Action/Button';

const Dropdown = ({ isToggled, actions }) => (
  <Card
    className={cn({
      [styles.Dropdown]: !isToggled,
      [styles.Dropdown___toggled]: isToggled,
    })}
  >
    {actions?.map((action, index) =>
      action.type === actionTypes.LINK ? (
        <ActionLink
          key={`Action-${index}`}
          action={action}
          data-test="actionLink"
        />
      ) : (
        <ActionButton
          key={`Action-${index}`}
          action={action}
          data-test="actionButton"
        />
      )
    )}
  </Card>
);

Dropdown.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.oneOf([actionTypes.BUTTON, actionTypes.LINK]),
      action: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
      icon: PropTypes.string,
      text: PropTypes.string,
      tooltip: PropTypes.string,
      disabled: PropTypes.bool,
    })
  ).isRequired,
  isToggled: PropTypes.bool.isRequired,
};

export default Dropdown;
