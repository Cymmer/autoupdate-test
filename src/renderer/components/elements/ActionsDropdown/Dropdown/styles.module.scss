@import '../../../../scss/variables';

@mixin base {
  position: absolute;
  z-index: 2;
  right: 0;
  padding: 0;
  margin: 0;
  min-width: 140px;
  border-radius: 8px;
  box-shadow: $shadow-raised;
  border: none;

  @content;
}

.Dropdown {
  @include base {
    display: none;
  }

  &___toggled {
    @include base;
  }
}
