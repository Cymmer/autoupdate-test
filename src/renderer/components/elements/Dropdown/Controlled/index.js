import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Select from 'react-select';
import GLOBALS from 'codechum-app-globals';
import RequiredField from 'components/elements/RequiredField';
import styles from '../styles.module.scss';

import Text from '../../Text';
import textTypes from '../../Text/constants/textTypes';

import { getIsNightMode } from '../../../../ducks';

import dropdownTypes from '../constants/dropdownTypes';
import determineStyles from '../styles/determineStyles';

import CustomOption from '../custom/Option';
import CustomSingleValue from '../custom/SingleValue';
import CustomValueContainer from '../custom/ValueContainer';

/* eslint-disable react/display-name */
const ControlledDropdown = ({
  options,
  isMulti,
  name,
  value,
  error,
  onChange,
  placeholder,
  isRequired,
  helperText,
  className,
  type,
  label,
  isNightMode,
  isClearable,
  disabled,
}) => (
  <div className={cn(className, styles.Dropdown)} id={name}>
    {label && type === dropdownTypes.SLIM && (
      <Text
        className={styles.Dropdown___slim_label}
        colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
        data-test="dropdownLabel"
        element="label"
        id="dropdownLabel"
        type={textTypes.BODY.MD}
      >
        {isRequired ? <RequiredField placeholder={label} /> : label}
      </Text>
    )}
    <Select
      components={{
        Option: CustomOption,
        SingleValue: CustomSingleValue,
        ValueContainer: (valueContainerProps) => (
          <CustomValueContainer type={type} {...valueContainerProps} />
        ),
        IndicatorSeparator: null,
      }}
      isClearable={isClearable}
      isDisabled={disabled}
      isMulti={isMulti}
      name={name}
      options={options}
      placeholder={
        isRequired ? <RequiredField placeholder={placeholder} /> : placeholder
      }
      styles={determineStyles(type, isNightMode)}
      value={value}
      onChange={onChange}
    />
    {(helperText || error) && (
      <div className={styles.Dropdown_helperTextContainer}>
        {helperText && (
          <Text
            className={styles.Dropdown_helperText}
            colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
            type={textTypes.BODY.XS}
          >
            {helperText}
          </Text>
        )}
        {error && (
          <Text
            colorClass={GLOBALS.COLOR_CLASSES.RED['500']}
            data-test="inputError"
            type={textTypes.BODY.XS}
          >
            {error}
          </Text>
        )}
      </div>
    )}
  </div>
);

ControlledDropdown.defaultProps = {
  className: null,
  label: null,
  type: dropdownTypes.FORM,
  value: null,
  isMulti: false,
  placeholder: null,
  helperText: null,
  error: null,
  isClearable: false,
  disabled: false,
};

ControlledDropdown.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string,
  error: PropTypes.string,
  type: PropTypes.oneOf([
    dropdownTypes.FORM,
    dropdownTypes.SLIM,
    dropdownTypes.PLAYGROUND,
    dropdownTypes.LARGE,
  ]),
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  options: PropTypes.oneOfType([
    // Proptype for the non-grouped options
    PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
          .isRequired,
        label: PropTypes.string.isRequired,

        // custom icon in each option (a custom component)
        icon: PropTypes.element,
      })
    ),

    // Proptype for the grouped options
    PropTypes.arrayOf(
      PropTypes.shape({
        label: PropTypes.string.isRequired,
        options: PropTypes.arrayOf(
          PropTypes.shape({
            value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
              .isRequired,
            label: PropTypes.string.isRequired,

            // custom icon in each option (a custom component)
            icon: PropTypes.element,
          })
        ),
      })
    ),
  ]).isRequired,
  isMulti: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.shape({
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
      label: PropTypes.string.isRequired,

      // custom icon in each option (a custom component)
      icon: PropTypes.element,
    }),
    PropTypes.arrayOf(
      PropTypes.shape({
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
          .isRequired,
        label: PropTypes.string.isRequired,

        // custom icon in each option (a custom component)
        icon: PropTypes.element,
      })
    ),
  ]),
  placeholder: PropTypes.string,
  helperText: PropTypes.string,
  isNightMode: PropTypes.bool.isRequired,
  isClearable: PropTypes.bool,
  disabled: PropTypes.bool,
  isRequired: PropTypes.bool,
};

const mapStateToProps = (store) => ({
  isNightMode: getIsNightMode(store),
});

export default connect(mapStateToProps, null)(ControlledDropdown);
