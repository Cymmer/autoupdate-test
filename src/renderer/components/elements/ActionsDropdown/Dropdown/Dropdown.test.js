import React from 'react';
import { shallow } from 'enzyme';
import Dropdown from './index';
import actionTypes from '../constants/actionTypes';

describe('RENDERING - ActionsDropdown > Dropdown', () => {
  it('should render with default props', () => {
    const component = shallow(<Dropdown actions={[]} isToggled={false} />);
    expect(component).toMatchSnapshot();
  });

  it('should render correct and appropriate children', () => {
    const component = shallow(
      <Dropdown
        actions={[
          {
            type: actionTypes.BUTTON,
            action: () => {},
            icon: 'test',
            text: 'Test',
            tooltip: 'Test Tooltip',
            disabled: false,
          },
          {
            type: actionTypes.BUTTON,
            action: () => {},
            icon: 'test',
            text: 'Test',
            tooltip: 'Test Tooltip',
            disabled: false,
          },
          {
            type: actionTypes.LINK,
            action: 'www.codechum.com',
            icon: 'test',
            text: 'Test',
            tooltip: 'Test Tooltip',
            disabled: false,
          },
          {
            type: actionTypes.LINK,
            action: 'www.codechum.com',
            icon: 'test',
            text: 'Test',
            tooltip: 'Test Tooltip',
            disabled: false,
          },
        ]}
        isToggled={false}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
