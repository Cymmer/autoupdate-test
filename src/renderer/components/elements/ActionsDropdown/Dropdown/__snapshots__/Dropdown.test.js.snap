// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDERING - ActionsDropdown > Dropdown should render correct and appropriate children 1`] = `
<Card
  className="Dropdown"
  id={null}
  onClick={null}
  onMouseEnter={null}
  onMouseLeave={null}
>
  <ActionButton
    action={
      Object {
        "action": [Function],
        "disabled": false,
        "icon": "test",
        "text": "Test",
        "tooltip": "Test Tooltip",
        "type": "button",
      }
    }
    data-test="actionButton"
    key="Action-0"
  />
  <ActionButton
    action={
      Object {
        "action": [Function],
        "disabled": false,
        "icon": "test",
        "text": "Test",
        "tooltip": "Test Tooltip",
        "type": "button",
      }
    }
    data-test="actionButton"
    key="Action-1"
  />
  <ActionLink
    action={
      Object {
        "action": "www.codechum.com",
        "disabled": false,
        "icon": "test",
        "text": "Test",
        "tooltip": "Test Tooltip",
        "type": "link",
      }
    }
    data-test="actionLink"
    key="Action-2"
  />
  <ActionLink
    action={
      Object {
        "action": "www.codechum.com",
        "disabled": false,
        "icon": "test",
        "text": "Test",
        "tooltip": "Test Tooltip",
        "type": "link",
      }
    }
    data-test="actionLink"
    key="Action-3"
  />
</Card>
`;

exports[`RENDERING - ActionsDropdown > Dropdown should render with default props 1`] = `
<Card
  className="Dropdown"
  id={null}
  onClick={null}
  onMouseEnter={null}
  onMouseLeave={null}
/>
`;
